Staging/adxl313 driver

danmois opened the pull request:
The implementation of the ADXL313 driver and the addition of the BASIC example for the ADXL313Z development board.

danmois commented:
The builds.json file was not added, and the project is not built using the automatic tool (analogdevicesinc.no-OS (Projects) test). The file will be added after the infrastructure for automatic checking of STM32 projects is updated.

CiprianRegus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+int adxl314_conf_inact_thr(struct adxl314_dev *dev, uint8_t inact_thr)
+{
+	int ret = -EINVAL;
+	uint8_t data[1] = {inact_thr};
```
Should this be an array?

CiprianRegus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+int adxl314_conf_act_thr(struct adxl314_dev *dev, uint8_t act_thr)
+{
+	int ret = -EINVAL;
+	uint8_t data[1] = {act_thr};
```
Should this be an array?

danmois replied:
Same as previous reply.

CiprianRegus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+		ret = adxl314_control_reg_bit(dev, ADXL314_REG_POWER_CTL,
+					      ENABLE_E, ADXL314_POWER_CTL_MEASURE);
+		break;
+	default:
+		pr_err("ADXL314 operation mode not supported!\n");
+		goto error_cmd;
+		break;
+	}
+
+	if (ret)
+		goto error_com;
+
+	dev->op_mode = op_mode;
+
+	return ret;
+error_com:
+	return -EPIPE;
+error_cmd:
+	return -ENOSYS;
+}
```
This can be simplified by removing the error labels and returning directly.

CiprianRegus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+
+	if (dev->comm_type == ADXL314_SPI_COMM) {
+		ret = no_os_spi_init(&dev->com_desc.spi_desc, &(init_param.comm_init.spi_init));
+		if (ret)
+			goto error_dev;
+	} else {
+		ret = no_os_i2c_init(&dev->com_desc.i2c_desc, &init_param.comm_init.i2c_init);
+		if (ret)
+			goto error_dev;
```
nit: These checks can be moved outside of the higher level if.

CiprianRegus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+	if (ret)
+		goto error_com;
+
+	pr_info("ADXL314 successfully initialized.\n");
```
I might be wrong, but I don't think there is a point in using  these `pr_info` messages, since we use error codes.

danmois replied:
The only ones that are left are concerned with the success of the self-test.

danmois commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+		reg_data |= modified_bit;
+
+	/* Write data to register for turning on desired operation mode. */
+	ret = adxl314_write_device_data(dev, base_address, 1, &reg_data);
+	if (ret)
+		return ret;
+
+	return ret;
```
Thanks! Solved.

danmois commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+int adxl314_write_device_data(struct adxl314_dev *dev, uint8_t base_address,
+			      uint16_t size, uint8_t *write_data)
+{
+	int ret = -ENXIO;
```
Thank you for the review! I performed the recommended modifications.

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+	else
+		*op_mode = ADXL314_STDBY;
+
+	return ret;
```
nit: return 0? you already checked if different than 0.

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+		return -EINVAL;
+	}
+
+	return ret;
```
ditto: return 0

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+		}
+	}
+
+	return ret;
```
ditto: return 0

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+			  struct adxl314_frac_repr *y,
+			  struct adxl314_frac_repr *z)
+{
+	int ret = -EINVAL;
```
ditto: no need for initialization

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+*******************************************************************************/
+int adxl314_conf_act_thr(struct adxl314_dev *dev, uint8_t act_thr)
+{
+	int ret = -EINVAL;
```
ditto: no need for initialization

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+*******************************************************************************/
+int adxl314_conf_inact_thr(struct adxl314_dev *dev, uint8_t inact_thr)
+{
+	int ret = -EINVAL;
```
same as above no initialization needed, check `ret` after `write_device_data` and return 0 at the end of the function. applies for most of the functions below.

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+	/* Set MEAS operation mode. */
+	ret = adxl314_set_op_mode(dev, ADXL314_MEAS);
+
+	return ret;
```
return directly `adxl314_set_op_mode`

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+		return ret;
+
+	/* Set MEAS operation mode. */
+	ret = adxl314_set_op_mode(dev, ADXL314_MEAS);
```
ditto: return directly

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+	int ret = -EINVAL;
+	uint8_t reg_val = en_ctl.value;
+
+	ret = adxl314_write_device_data(dev, ADXL314_REG_INT_EN, 1, &reg_val);
```
ditto: return directly. Please check all functions for which this statement applies

amiclaus commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1146 @@
+	accel_data = (int16_t)raw_accel;
+
+	/* Apply scale factor based on the selected range. */
+	return ((int64_t)(accel_data * ADXL314_ACC_SCALE_FACTOR_MUL));
```
is this cast really required? (accel_data is already int64_t)

danmois replied:
You are right, there is no need for the cast here.

amiclaus reviewed (commented):
some comments on my side.

danmois commented:
V1

- function return value initialization removal
- removal of UART messages from the initialization function

danmois commented on projects/eval-adxl314z/src/examples/iio_example/iio_example.c:
```diff
@@ -0,0 +1,91 @@
+/******************************************************************************/
+/************************ Variable Declarations ******************************/
+/******************************************************************************/
+uint8_t iio_data_buffer[DATA_BUFFER_SIZE*3*sizeof(int)];
```
Sorry, these files should not have been included in the PR.

danmois commented on projects/eval-adxl314z/src/examples/iio_trigger_example/iio_trigger_example.c:
```diff
@@ -0,0 +1,123 @@
+/******************************************************************************/
+/************************ Variable Declarations ******************************/
+/******************************************************************************/
+uint8_t iio_data_buffer[DATA_BUFFER_SIZE*3*sizeof(int)];
```
Sorry, these files should not have been included in the PR.

danmois commented:
V2

- removed redundant code regarding function return values
- removed unnecessary cast in adxl314_accel_conv function
- removed empty lines

rbolboac commented on projects/eval-adxl314z/src/platform/stm32/main.c:
```diff
@@ -0,0 +1,108 @@
+
+	stm32_init();
+
+#ifdef IIO_EXAMPLE
+	ret = iio_example_main();
+	if (ret < 0)
+		goto error;
+#endif
+
+#ifdef IIO_TRIGGER_EXAMPLE
+	ret = HAL_EXTI_SetConfigLine(&xiip, &adxl355_int_exticonfig);
+	if (ret != HAL_OK)
+		return ret;
+	ret = iio_trigger_example_main();
+	if (ret < 0)
+		goto error;
+#endif
```
If these examples are not yet supported I would add an error to inform the user:
`#error Selected example is not supported.`

danmois replied:
Code deleted until examples implemented.

rbolboac commented on projects/eval-adxl314z/Makefile:
```diff
@@ -0,0 +1,10 @@
+# Select the example you want to enable by chossing y for enabling and n for disabling
+DUMMY_EXAMPLE = y
+IIO_EXAMPLE = n
+IIO_TRIGGER_EXAMPLE = n
```
nitpick: I would remove these if the examples are not yet supported, to avoid confusion.

rbolboac commented on drivers/accel/adxl314/adxl314.h:
```diff
@@ -0,0 +1,483 @@
+
+/* ADXL314 register map */
+#define ADXL314_DEVID_AD 	 			0x00
+#define ADXL314_REG_OFS_AXIS(index)		(0x1E + (index))
```
nitpick: please check indentation throughout the file

rbolboac commented on drivers/accel/adxl314/adxl314.h:
```diff
@@ -0,0 +1,483 @@
+/**
+ * @enum adxl314_op_mode
+ * @brief Enum for operating mode.
+ */enum adxl314_op_mode {
```
nitpick: add enum on a new line

rbolboac commented on drivers/accel/adxl314/adxl314.h:
```diff
@@ -0,0 +1,483 @@
+/**
+ * @enum adxl314_lp_mode
+ * @brief Enum for low power operating mode.
+ */enum adxl314_lp_mode {
```
nitpick: add enum on a new line

rbolboac commented on drivers/accel/adxl314/adxl314.h:
```diff
@@ -0,0 +1,483 @@
+	/** Device communication descriptor */
+	union adxl314_comm_desc com_desc;
+	/** Device Communication type: ADXL314_SPI_COMM, ADXL314_I2C_COMM */
+	enum adxl314_comm_type comm_type;
+	enum adxl314_op_mode op_mode;
+	enum adxl314_lp_mode lp_mode;
+	enum adxl314_odr odr;
+	uint16_t x_offset;
+	uint16_t y_offset;
+	uint16_t z_offset;
+	enum adxl314_fifo_mode fifo_mode;
+	uint8_t fifo_samples;
+	uint8_t act_thr;
+	uint8_t inact_thr;
+	uint8_t time_inact;
+	union adxl314_act_inact_ctl_flags act_inact_ctl;
+	uint8_t comm_buff[24];
```
nitpick: add doxy comment for each member

rbolboac commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1119 @@
+}
+
+/*******************************************************************************
+ * @brief Set offset for each axis. Scale factor of 1.56 g/LSB.
```
nitpick: should it also be mentioned that the offset is in 2's complement?

rbolboac commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1119 @@
+	int16_t zst = 0;
+	uint8_t fifo_entries = 0;
+
+	/* 1. Set the device ODR to 200 Hz. */
```
the comment mentions a frequency of 200 Hz but 100 Hz is set

rbolboac commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1119 @@
+	if (ret)
+		return ret;
+
+	/* Result is 13 bits long with sign extended. */
```
is it 13-bit or 16-bit?

danmois replied:
It is a 13-bit result with sign extension.

danmois commented on projects/eval-adxl314z/src/platform/stm32/main.c:
```diff
@@ -0,0 +1,108 @@
+		goto error;
+#endif
+
+#if (IIO_EXAMPLE+IIO_TRIGGER_EXAMPLE+DUMMY_EXAMPLE != 1)
+#error Selected example projects cannot be enabled at the same time. \
+Please enable only one example and re-build the project.
+#endif
```
I would delete it entirely until supported. This should be done in the following days.

danmois commented:
V3

- removed code belonging to examples not yet supported
- modified alignment
- added Doxygen comments to ADXL314 device structure members

danmois commented:
V4

- fixed driver header file define alignment

danmois commented:
V5

- fixed return value for dummy example (ret used uninitialized)

adrimbarean commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1120 @@
+ * @return ret         - Result of the reading operation (0 - success, negative
+ *                       value for failure).
+*******************************************************************************/
+int adxl314_read_device_data(struct adxl314_dev *dev, uint8_t base_address,
```
I think we should add the `adxl314_reg_read(struct adxl314_dev *dev, uint8_t base_address, uint8_t *read_data)` and `adxl314_reg_write(struct adxl314_dev *dev, uint8_t base_address, uint8_t write_data)` functions. I think we find them more often in other drivers, maybe creating a recognizable pattern, as well as they will be useful for when we add the IIO driver.

adrimbarean commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1120 @@
+ *
+ * @return ret - Result of the operation (0 - success, negative value - failure).
+*******************************************************************************/
+int adxl314_control_reg_bit(struct adxl314_dev *dev,
```
Can this be accomplished as well with `adxl314_reg_write_msk()`.

danmois replied:
Yes, it can. I will remove the adxl314_control_reg_bit function.

adrimbarean commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1120 @@
+	if (ret)
+		return ret;
+
+	/* Clear data. */
```
nitpick: comments not necessary here ;) The code is clear enough.

adrimbarean commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1120 @@
+	uint8_t reg_value = 0;
+
+	dev = (struct adxl314_dev *)calloc(1, sizeof(*dev));
+
```
nitpick: extra line

adrimbarean commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1120 @@
+	dev->act_inact_ctl.value = reg_value;
+
+	/* Put device in Measure mode. */
+	ret = adxl314_set_op_mode(dev, ADXL314_MEAS);
+	if (ret)
+		goto error_com;
+
+	return ret;
```
We can return directly here.

adrimbarean commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1120 @@
+	else
+		ret = no_os_i2c_remove(dev->com_desc.i2c_desc);
+
+	free(dev);
```
We should only `free(dev)` after we're sure i2c/spi_remove is successful.

danmois commented:
V6

- removed the adxl314_control_reg_bit function
- used enum for axes selection in adxl314_read_device_data calls
- call free() if i2c/spi_remove is successful in adxl314_remove function

danmois commented:
V7

- changed user select-able offset from uint16_t to uint8_t

buha commented on projects/eval-adxl314z/src/examples/dummy/dummy_example.c:
```diff
@@ -0,0 +1,175 @@
+ * @return ret - Result of the example execution. If working correctly, will
+ *               execute continuously the while(1) loop and will not return.
+*******************************************************************************/
+int dummy_example_main()
```
rename `dummy_example` to something more meaningful like `basic` or `console_example`

buha commented on drivers/accel/adxl314/adxl314.c:
```diff
@@ -0,0 +1,1097 @@
+ * @return ret         - Result of the writing operation (0 - success, negative
+ *                       value for failure).
+*******************************************************************************/
+int adxl314_write_device_data(struct adxl314_dev *dev, uint8_t base_address,
```
rename to `adxl314_write` ? no need for that long name

buha commented on projects/eval-adxl314z/src/examples/dummy/dummy_example.c:
```diff
@@ -0,0 +1,175 @@
+		goto error;
+
+	/* Set offset for Z axis. */
+	ret = adxl314_set_offset(adxl314, 6, ADXL314_Z_AXIS);
```
this example should result in measuring ~9.8 on each axis without any offset

danmois replied:
The outputs on the three axes without applying any offset are within the values specified by the datasheet. The range of the device is between +/- 200g and that is why the values obtained seem to be far from the actual measured ones.

danmois commented:
V8

- renamed the adxl314_write_device_data and adxl314_read_device_data functions to adxl314_write and adxl314_read, respectively
- modified the set offset function to get offset in mg
- added the get offset function
- simplified the adxl314_accel_conv function
- renamed dummy_example to basic_example

danmois commented:
V9

- rebase master

danmois commented:
V8

- fixed data display alignment in the basic example

danmois commented:
V10
- Modification of functions for setting thresholds to accept settings in mg  
- Addition of functions for getting threshold values

CiprianRegus commented on projects/eval-adxl313z/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,175 @@
+	struct adxl313_dev *adxl313;
+	int ret = -EINVAL;
+
+	/* Clear the screen. */
+	pr_info("%c",27);
+	pr_info("%c",'[');
+	pr_info("%c",'2');
+	pr_info("%c",'J');
```
Does this work for any terminal?

danmois replied:
I think it should.

danmois commented:
V11

- driver modifications for including all ADXL313 features

danmois commented on projects/eval-adxl313z/src/common/common_data.c:
```diff
@@ -0,0 +1,88 @@
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#ifdef BASIC_EXAMPLE
```
I changed the naming after a previous discussion: https://github.com/analogdevicesinc/no-OS/pull/1356#discussion_r864712275. The conclusion was that this naming is more appropriate. However, changes can still be performed.

danmois commented on projects/eval-adxl313z/src/common/common_data.c:
```diff
@@ -0,0 +1,88 @@
+};
+#endif
+
+#ifdef IIO_TRIGGER_EXAMPLE
+struct no_os_irq_init_param adxl313_int_ip = {
+	.irq_ctrl_id = ADXL313_TRIGGER_INTR_ID,
+	.platform_ops = IRQ_OPS,
+	.extra = &xiip,
+};
+
+struct adxl313_iio_trig_init_param adxl313_iio_trig_user_init = {
+	.irq_init_param = &adxl313_int_ip,
+	.name = IIO_ADXL313_TRIGGER_NAME,
+};
+
+struct adxl313_iio_sw_trig_init_param adxl313_iio_sw_trig_user_init = {
+	.name = IIO_ADXL313_SW_TRIGGER_NAME,
+};
+#endif
```
Thank you for the observation, I removed the section.

danmois commented:
V12

- added MACRO for baud rate and SPI frequency
- removed IIO_TRIGGER example

danmois commented:
V13

- addred the routines for activity/inactivity detection

danmois commented:
V14

- added functions for enabling Link Mod for activity and inactivity detection

danmois commented:
V14

- added support for ADXL312

danmois commented:
V15

- fixed comment typos

danmois commented:
V16

- set proper notation for offset in function for setting/reading offset value

danmois commented:
V17

- fixed ODR bug in device initialization function

danmois commented:
V18

- removed IIO_TRIGGER_EXAMPLE code

rbolboac commented on projects/eval-adxl313z/src/examples/examples_src.mk:
```diff
@@ -0,0 +1,27 @@
+ifeq (y,$(strip $(IIO_EXAMPLE)))
+TINYIIOD=y
+CFLAGS += -DIIO_EXAMPLE=1
+SRCS += $(PROJECT)/src/examples/iio_example/iio_example.c
+INCS += $(PROJECT)/src/examples/iio_example/iio_example.h
+endif
+
+ifeq (y,$(strip $(IIO_TRIGGER_EXAMPLE)))
+TINYIIOD=y
+CFLAGS += -DIIO_TRIGGER_EXAMPLE=1
+SRCS += $(PROJECT)/src/examples/iio_trigger_example/iio_trigger_example.c
+INCS += $(PROJECT)/src/examples/iio_trigger_example/iio_trigger_example.h
+endif
```
nitpick: remove these lines until the examples are available.

danmois commented on projects/eval-adxl313z/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,216 @@
+
+		/* Read FIFO data. */
+		ret = adxl313_get_fifo_data(adxl313, &fifo_entries, x, y, z);
+		//ret = adxl313_get_raw_fifo_data(adxl313, &fifo_entries, x_raw_array, y_raw_array, z_raw_array);
```
It was unintentionally left. Thank you for the comment, I will remove the line!

amiclaus reviewed (commented):
I believe the .json can be added now, after a rebase with the latest master branch. (d2018ba)

danmois commented:
V19

- removed code for unused examples
- added the builds.json file

danmois commented:
V20

- rebase

adrimbarean reviewed (approved):
Good and needed addition. Seems ready to merge.

danmois commented:
V21

- added self-test validation for ADXL312
- bug fix for range setting for ADXL312
- modified basic_example so that ADXL312 is also supported

buha commented:
i am merging this into master as it was tested on various chips from this family and it works
